File: packages/newton-dsl-py/src/newton/edges.py
"""
Edge / transition helpers.

Transitions are added via Task.then() and Task.otherwise().
Priority is determined by declaration order:
  - First .then() call gets priority 0
  - Next gets priority 5, 10, 15, ...
  - .otherwise() always gets lowest priority (no `when` condition)
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .refs import Guard, Ref, when as _when

if TYPE_CHECKING:
    pass


PRIORITY_STEP = 5


class EdgeSpec:
    """
    Internal representation of a single transition edge before compilation.
    """

    def __init__(
        self,
        target_id: str,
        *,
        priority: int,
        guard: Guard | None = None,
        label: str | None = None,
    ) -> None:
        self.target_id = target_id
        self.priority = priority
        self.guard = guard
        self.label = label

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"to": self.target_id}
        if self.priority != 100:
            d["priority"] = self.priority
        if self.guard is not None:
            d["when"] = self.guard.to_condition()
        if self.label is not None:
            d["label"] = self.label
        return d

File: packages/newton-dsl-ts/src/edges.ts
/**
 * Edge / transition helpers.
 *
 * Transitions are added via Task.then() and Task.otherwise().
 * Priority is determined by declaration order:
 *   - First .then() call gets priority 0
 *   - Next gets priority 5, 10, 15, ...
 *   - .otherwise() always uses current priority (no guard), then advances
 */

import type { Guard } from "./refs.js";

export const PRIORITY_STEP = 5;

export interface EdgeDict {
  to: string;
  priority?: number;
  when?: { $expr: string };
  label?: string;
}

/**
 * Internal representation of a single transition edge before compilation.
 */
export class EdgeSpec {
  constructor(
    public readonly targetId: string,
    public readonly priority: number,
    public readonly guard: Guard | null = null,
    public readonly label: string | null = null
  ) {}

  toDict(): EdgeDict {
    const d: EdgeDict = { to: this.targetId };
    if (this.priority !== 100) {
      d.priority = this.priority;
    }
    if (this.guard !== null) {
      d.when = this.guard.toCondition();
    }
    if (this.label !== null) {
      d.label = this.label;
    }
    return d;
  }
}
